src/v/base/vassert.h

```diff
@@ -17,11 +17,64 @@
 #include <seastar/util/backtrace.hh>
 #include <seastar/util/log.hh>
 
+#include <atomic>
+
 namespace detail {
 struct dummyassert {
     static inline ss::logger l{"assert"};
 };
 inline dummyassert g_assert_log;
+/**
+ * @brief Class used to hold messages generated by asserts
+ *
+ * This class holds messages generated by asserts which will be fetched by the
+ * crash tracker when handling the signal generated by `vassert`.
+ */
+class assert_log_holder {
+public:
+    /**
+     * @brief Registers a vassert event
+     *
+     * @tparam Args The argument template
+     * @param bt The backtrace from the assert
+     * @param fmt The format string for the log
+     * @param args The arguments to @p fmt
+     */
+    template<typename... Args>
+    void register_event(
+      ss::saved_backtrace bt,
+      ss::logger::format_info_t<Args...> fmt,
+      Args&&... args) noexcept {
+        if (event_registered()) {
+            // only register one event per thread
+            return;
+        }
+
+        _buffer = fmt::format(
+          fmt::runtime(fmt.format), std::forward<Args>(args)...);
+        _abuffer.store(_buffer.data());
+
+        g_assert_log.l.error("{}", _buffer);
+        g_assert_log.l.error("Backtrace:\n{}", bt);
+    }
+
+    const ss::sstring& held_event() const noexcept { return _buffer; }
+    bool event_registered() const noexcept { return !_buffer.empty(); }
+    bool event_registered_signal_safe() const noexcept {
+        return _abuffer.load() != nullptr;
+    }
+    std::string_view get_signal_safe_log() noexcept {
+        return {_abuffer.load()};
+    }
+
+private:
+    ss::sstring _buffer;
+    // According to https://en.cppreference.com/w/c/program/signal, referring to
+    // objects with static or thread-local storage duration that is not
+    // lock-free atomic is UB.
+    static thread_local inline std::atomic<const char*> _abuffer{nullptr};
+};
+static thread_local assert_log_holder g_assert_log_holder;
```

Comment:
i'm concerned this isn't going to do what you want (or that i'm mistaken about what the exact goal is). here you're going to i think get a separate instance of `assert_log_holder` in each translation unit (`static foo` in a header), then those would all contain another thread_local `_abuffer` that points to a single instance (since its inline)? it's a bit of brain twister.

would this be simpler:

```
class assert_log_holder {
  std::atomic...;
};

inline thread_local assert_log_holder g_assert_log_holder
```

here you get a single global instance of `assert_log_holder` per thread.